tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/lhlo_ops.cc

```diff
@@ -226,6 +257,24 @@ static LogicalResult Verify(CustomCallOp op) {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// MapOp
+//===----------------------------------------------------------------------===//
+
+void MapOp::getSuccessorRegions(Optional<unsigned> index,
```

Comment:
I don't think the region attached to Map is the successor of the operation (as defined [here](https://github.com/llvm/llvm-project/blob/92880ab7a2b2145f0605f367cd6d53d6892903c3/mlir/include/mlir/Interfaces/ControlFlowInterfaces.td#L130)).  Rather, is part of the operation definition, like convolution dim numbers are for a convolution.

I'm new to this so please push back if I'm not making sense.

Reply:
That is indeed an interesting question. I think the original modelling of these interfaces had classical conditionals and loops in mind and that also shows in the doc comments.

However, I still think this is modeled correctly here. The body of the map operation has to execute before the map operation can complete. In that sense, the body region is the successor of the outer map operation. This does not cover the fact that it potentially runs multiple times in parallel etc. and that it has an implicit selection of elements from the operands.

An alternative would be to declare the body region as `IsolatedFromAbove` to give it more function-like semantics. That would in essence turn the map body into its own scope for allocation. I don't think that is supported, though. Currently, none of this matters as the map function is on scalars and hence typically does not allocate.

The same is true for the other ops, except for while, case, cond and fusion op, where the branch op interface is the correct choice.

Reply:
I don't quite get why the region isn't the successor of the operation? What is the predecessor of the region here then?

Reply:
@joker-eph I have the same question regarding the predecessor of this region.